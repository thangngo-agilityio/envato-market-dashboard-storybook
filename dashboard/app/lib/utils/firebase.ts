import {
  Timestamp,
  arrayUnion,
  doc,
  getFirestore,
  onSnapshot,
  serverTimestamp,
  setDoc,
  updateDoc,
} from 'firebase/firestore';
import { TMessages } from '../interfaces';
import { Messaging, getToken } from 'firebase/messaging';
import { initializeApp } from 'firebase/app';
import { getStorage } from 'firebase/storage';
import { getAuth } from 'firebase/auth';
import { FIREBASE_CHAT, USER_CHATS_FIELD } from '../constants';

const firebaseConfig = {
  apiKey: process.env.NEXT_PUBLIC_FIREBASE_API_KEY,
  authDomain: process.env.NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN,
  projectId: process.env.NEXT_PUBLIC_FIREBASE_PROJECT_ID,
  storageBucket: process.env.NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.NEXT_PUBLIC_FIREBASE_SENDER_ID,
  appId: process.env.NEXT_PUBLIC_FIREBASE_APP_ID,
  measurementId: process.env.NEXT_PUBLIC_FIREBASE_MEASUREMENT_ID,
};

initializeApp(firebaseConfig);

const VAPID_KEY = process.env.NEXT_PUBLIC_FIREBASE_VAPID;

export const requestForToken = async (messaging: Messaging) => {
  try {
    const currentToken = await getToken(messaging, { vapidKey: VAPID_KEY });
    if (currentToken) {
      return currentToken;
      // Perform any other neccessary action with the token
    } else {
      // Show permission request UI
      console.log(
        'No registration token available. Request permission to generate one.',
      );
    }
  } catch (err) {
    console.log('An error occurred while retrieving token. ', err);
  }
};

// Initialize Firebase
export const app = initializeApp(firebaseConfig);
export const auth = getAuth(app);
export const storage = getStorage();
export const db = getFirestore();

export const sendMessage = async (
  data: TMessages,
  idRoomChat: string,
  senderId: string,
  adminId: string,
  avatarUrl: string,
  avatarAdminUrl: string,
  displayName: string,
) => {
  // Update 'chats' document with the new message
  await updateDoc(doc(db, FIREBASE_CHAT.CHATS, idRoomChat), {
    messages: arrayUnion({
      id: data.id,
      text: data.text,
      senderId: data.senderId,
      date: Timestamp.now(),
    }),
  });

  onSnapshot(doc(db, FIREBASE_CHAT.USER_CHATS, adminId), (doc) => {
    if (!doc.exists()) {
      createUserChat(
        adminId,
        senderId,
        idRoomChat,
        data.text,
        avatarUrl,
        avatarAdminUrl,
        displayName,
      );
    }
  });

  await updateDoc(doc(db, FIREBASE_CHAT.USER_CHATS, adminId), {
    [idRoomChat]: {
      lastMessage: {
        text: data.text,
      },
      date: serverTimestamp(),
      userInfo: {
        uid: senderId,
        avatarUrl,
        avatarAdminUrl,
        displayName,
      },
    },
  });
};

export const adminSendMessage = async (
  data: TMessages,
  idRoomChat: string,
  adminId: string,
) => {
  await updateDoc(doc(db, FIREBASE_CHAT.CHATS, idRoomChat), {
    messages: arrayUnion({
      id: data.id,
      text: data.text,
      senderId: data.senderId,
      date: Timestamp.now(),
    }),
  });

  await updateDoc(doc(db, FIREBASE_CHAT.USER_CHATS, adminId), {
    [idRoomChat + USER_CHATS_FIELD.LAST_MESSAGE]: {
      text: data.text,
    },
    [idRoomChat + USER_CHATS_FIELD.DATE]: serverTimestamp(),
  });
};

const createUserChat = async (
  adminId: string,
  senderId: string,
  idRoomChat: string,
  text: string,
  avatarUrl: string,
  avatarAdminUrl: string,
  displayName: string,
) => {
  setDoc(doc(db, FIREBASE_CHAT.USER_CHATS, adminId), {
    [idRoomChat]: {
      lastMessage: {
        text: text,
      },
      date: serverTimestamp(),
      userInfo: {
        uid: senderId,
        avatarUrl,
        avatarAdminUrl,
        displayName,
      },
    },
  });
};
