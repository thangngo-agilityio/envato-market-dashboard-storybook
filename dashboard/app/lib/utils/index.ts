export * from './transaction';
export * from './product';
export * from './time';
export * from './sort';
export * from './user';
export * from './url';
export * from './helpers';
export * from './text';
export * from './localStorage';
export * from './image';
export * from './prefetch';
export * from './toast';
export * from './getErrorMessage';
export * from './firebase';
export * from './recentActivities';
export * from './logAcitvity';
