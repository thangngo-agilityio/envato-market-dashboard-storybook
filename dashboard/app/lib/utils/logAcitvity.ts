import { AxiosInstance } from 'axios';

// Service
import { recentActivitiesHttpService } from '../services';

// Interface
import { TRecentActivities } from '../interfaces';

// Constants
import { END_POINTS } from '../constants';

export const logActivity = (
  httpService: AxiosInstance,
  actionName: string,
  userId?: string,
) => {
  const interceptor = httpService.interceptors.response.use(
    async (response) => {
      const id = response.data['_id'];
      await recentActivitiesHttpService.post<TRecentActivities>(
        END_POINTS.RECENT_ACTIVITIES,
        {
          userId: id || userId,
          actionName: actionName,
        },
      );

      return response;
    },
  );

  return interceptor;
};
