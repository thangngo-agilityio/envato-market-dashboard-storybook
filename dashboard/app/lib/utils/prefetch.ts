import { QueryClient } from '@tanstack/react-query';
import { getStatistical } from '../services';

export const prefetchStatistical = async <T>(
  endPoint: string,
  queryClient: QueryClient,
) => {
  await queryClient.prefetchQuery({
    queryKey: [endPoint],
    queryFn: () => getStatistical<T>(endPoint),
  });
};
