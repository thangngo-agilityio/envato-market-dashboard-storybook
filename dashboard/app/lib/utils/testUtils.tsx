import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

const queryClient = new QueryClient();

export const renderQueryProviderTest = (component: JSX.Element) => {
  const { render } = testLibReactUtils;
  return render(
    <QueryClientProvider client={queryClient}>{component}</QueryClientProvider>,
  );
};
