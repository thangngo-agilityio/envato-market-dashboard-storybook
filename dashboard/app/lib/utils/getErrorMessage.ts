import { ERROR_MESSAGES } from '@/lib/constants';
import { AxiosError } from 'axios';

type ErrorResponse = {
  message: string;
};

export const getErrorMessageFromAxiosError = <T>(
  error: AxiosError<T>,
  defaultErrorMessage?: string,
): string => {
  const { response } = error;

  if (response?.data) {
    const errorData = response.data as unknown as ErrorResponse;
    return errorData.message;
  }

  return defaultErrorMessage ?? ERROR_MESSAGES.DEFAULT_ERROR;
};
