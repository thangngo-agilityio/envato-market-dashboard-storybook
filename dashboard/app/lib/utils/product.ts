import dayjs from 'dayjs';

// Types
import { TProduct } from '@/lib/interfaces';

// Utils
import {
  formatAmountNumber,
  formatDecimalNumber,
  formatUppercaseFirstLetter,
} from '.';

// Constants
import { IMAGES, PRODUCT_STATUS, TIME_FORMAT } from '../constants';

/**
 * Convert data show for home page
 * @param products
 * @returns
 */
export const formatProductResponse = (products: TProduct[] = []) =>
  products.map((product) => {
    const {
      _id,
      name,
      amount,
      currency,
      createdAt,
      description,
      imageURLs,
      stock,
    } = product;

    return {
      _id: _id,
      name: formatUppercaseFirstLetter(name),
      description: formatUppercaseFirstLetter(description),
      date: dayjs(createdAt).format(TIME_FORMAT),
      imageURLs: imageURLs || IMAGES.SIGN_UP.url,
      amount: `${currency}${formatDecimalNumber(+amount)}`,
      stock: `${formatAmountNumber(stock.toString())}`,
      productStatus: +stock > 0 ? PRODUCT_STATUS.IN_STOCK : PRODUCT_STATUS.SOLD,
      product: { ...product },
    };
  });
