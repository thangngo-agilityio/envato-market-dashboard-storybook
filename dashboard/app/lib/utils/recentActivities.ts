import dayjs from 'dayjs';

// Types
import { TRecentActivities } from '@/lib/interfaces';

// Utils
import { formatUppercaseFirstLetter } from '.';

// Constants
import { TIME_DETAIL_FORMAT } from '../constants';

/**
 * Convert data show for home page
 * @param recentActivities
 * @returns
 */
export const formatRecentActivitiesResponse = (
  recentActivities: TRecentActivities[] = [],
) =>
  recentActivities.map((activity) => {
    const { _id, actionName, email, createdAt } = activity;

    return {
      _id: _id,
      actionName: formatUppercaseFirstLetter(actionName),
      email,
      date: dayjs(createdAt).format(TIME_DETAIL_FORMAT),
    };
  });
