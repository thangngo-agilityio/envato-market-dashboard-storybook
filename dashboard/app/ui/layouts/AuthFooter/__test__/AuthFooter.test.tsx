import '@testing-library/jest-dom';

// component
import AuthFooterComponent from '..';

// Utils
import { renderQueryProviderTest } from '@/lib/utils/testUtils';

describe('Auth footer component render', () => {
  const renderComponent = () =>
    renderQueryProviderTest(<AuthFooterComponent />);

  it('Should render match with snapshot.', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });
});
