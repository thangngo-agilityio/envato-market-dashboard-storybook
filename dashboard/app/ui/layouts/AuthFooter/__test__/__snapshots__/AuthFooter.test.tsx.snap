// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth footer component render Should render match with snapshot. 1`] = `
<div>
  <div
    class="css-1dfptg7"
  >
    <p
      class="chakra-text css-1k9cp2o"
    >
      @ 
      2024
       Bankco. All Rights Reserved.
    </p>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
